ap/tasks/reduction.py

```diff
@@ -220,28 +240,91 @@ def get_avg_std(values):
 )
 
 
-class MergeReducedEvents(DatasetTask, CalibratorsSelectorMixin, law.tasks.ForestMerge, HTCondorWorkflow):
-
-    sandbox = dev_sandbox("bash::$AP_BASE/sandboxes/venv_columnar.sh")
-
-    shifts = {SelectEvents}
+class MergeReducedEventsUser(DatasetTask):
 
     # recursively merge 8 files into one
     merge_factor = 8
 
-    # the key of the config that defines the merging in the branch map of DatasetTask
-    file_merging = "after_reduction"
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # cached value of the file_merging until it's positive
+        self._cached_file_merging = -1
+
+        # in case this is a workflow, do not cache branches by default
+        # (this is enabled in reduced_file_merging once positive)
+        self._cache_branches = False
+
+    @property
+    def file_merging(self):
+        """
+        Needed by DatasetTask to define the default branch map.
+        """
+        if self._cached_file_merging < 0:
+            # reset the forest in case this is a forest ForestMerge task
+            self._forest_built = False
```

Comment:
I just searched where the forest built is activated (since I assume this is the switch which activates it), but I couldn't find it here. I'm relatively certain though that the forest merge is activated at some point. Can you point me to the piece of code where this is done?

Reply:
Yep, this happens implicitly after the merge forest is build within the [law ForestMerge contrib task](https://github.com/riga/law/blob/master/law/contrib/tasks/__init__.py#L311), but only if caching is enabled (the line above the referenced one). It's up to the tasks to decide whether they use caching or not. One example is `MergeHistograms.merge_workflow_requires` in this PR.